core/src/main/java/com/graphhopper/GraphHopper.java

```diff
@@ -797,31 +857,22 @@ public final CHPreparationHandler getCHPreparationHandler() {
     }
 
     private void initCHPreparationHandler() {
-        if (!chPreparationHandler.hasCHProfiles()) {
-            if (chPreparationHandler.getCHProfileStrings().isEmpty())
-                throw new IllegalStateException("Potential bug: chProfileStrings is empty");
-
-            for (FlagEncoder encoder : encodingManager.fetchEdgeEncoders()) {
-                for (String chWeightingStr : chPreparationHandler.getCHProfileStrings()) {
-                    // ghStorage is null at this point
-
-                    // extract weighting string and u-turn-costs
-                    String configStr = "";
-                    if (chWeightingStr.contains("|")) {
-                        configStr = chWeightingStr;
-                        chWeightingStr = chWeightingStr.split("\\|")[0];
-                    }
-                    PMap config = new PMap(configStr);
-                    int uTurnCosts = config.getInt(Routing.U_TURN_COSTS, INFINITE_U_TURN_COSTS);
-
-                    EdgeBasedCHMode edgeBasedCHMode = chPreparationHandler.getEdgeBasedCHMode();
-                    if (!(edgeBasedCHMode == EDGE_OR_NODE && encoder.supportsTurnCosts())) {
-                        chPreparationHandler.addCHProfile(CHProfile.nodeBased(createWeighting(new HintsMap(chWeightingStr), encoder, NO_TURN_COST_PROVIDER)));
-                    }
-                    if (edgeBasedCHMode != OFF && encoder.supportsTurnCosts()) {
-                        chPreparationHandler.addCHProfile(CHProfile.edgeBased(createWeighting(new HintsMap(chWeightingStr), encoder, new DefaultTurnCostProvider(encoder, ghStorage.getTurnCostStorage(), uTurnCosts))));
-                    }
-                }
+        if (chPreparationHandler.hasCHProfiles()) {
+            return;
+        }
+
+        Map<String, ProfileConfig> profilesByName = getProfilesByName();
+        for (CHProfileConfig chConfig : chPreparationHandler.getCHProfileConfigs()) {
+            ProfileConfig profile = profilesByName.get(chConfig.getProfile());
+            assert profile != null : "there is no profile " + chConfig.getProfile();
```

Comment:
The reason I am using assert here is that this is a private method and the existence of the profile is already checked in `checkProfilesConsistency` before. So at this point its clear that the profile exists (unless there is some programming error and my assumption that it was checked before is false). So we could either skip this check here entirely, or repeat the check again, or use an assert as a bit of a reminder that (if nothing is broken) we can be sure the profile exists. In general of course you are right and the program logic should not depend on an assert. Here it does not because the same check is done before already. If you think this is confusing somehow I think we should just remove this assert here (and use the null pointer exception to figure out what is going on if this ever pops up). I have to admit the assert message here is not chosen very well it would better be: "There is no profile ... even though this should have been checked before"

Reply:
An assert statement always distracts a bit and should have some reason it is no if+throw because even if there is a bug we'll never know with assert, because in production it is disabled and in tests it does not happen. (except we create some randomization tests that randomize the initialization too)

So for higher level code where speed does not matter I would always use an if statement + throw an IllegalStateException OR decide it is not worth the clutter and remove it.

Interestingly even for code where speed matters I observed that "if+throw" can often be removed via JIT and does not hurt. E.g. recently I introduced this in RAMDataAccess.setInt and I should probably replace the remaining assert too or remove it.

Reply:
> An assert statement always distracts a bit

Thats a bit subjective :) I think asserts can do a great job in communicating intend (even though the present case is not a particularly good example for this).

>  it is no if+throw because even if there is a bug we'll never know with assert, because in production it is disabled and in tests it does not happen

Yes and like I said I did not add the assert because its supposed to throw anything (we check wether or not the profile exists in `checkConsistency()` already.

> So for higher level code where speed does not matter I would always use an if statement

I did not decide to use assert for performance reasons here. Performing a single if is probably very rarely a performance issue (and especially not here).

What do you prefer?
```java
// option 1:
Profile profile = profiles.get(profileName);
// profile cannot be null because we checked already
profile.doSth();

// option2
Profile profile = profiles.get(profile);
profile.doSth();

// option3
Profile profile = profiles.get(profile);
assert profile != null : "profile does not exist even though this should have been checked already";
profile.doSth();
```

Here, I am happy with any of the options, but I would not agree if we tried to never use assert whatsoever. Like I said this is not a particular good/important use-case here.

Reply:
If I'm confident that it is really not null I would probably use option 2.

If not, I would pick option 4: throw an IllegalStateException ;)
I still do not see the benefit of an assert that is never thrown in production. Why should I continue in production when we are in a wrong situation? It feels worse than catching an exception and then just logging it :)

Reply:
> If I'm confident that it is really not null I would probably use option 2.

Then the next person might ask the same question and maybe he/she is not as confident :)

> If not, I would pick option 4: throw an IllegalStateException ;)

Ok that also works for me.

> I still do not see the benefit of an assert that is never thrown in production. Why should I continue in production when we are in a wrong situation? It feels worse than catching an exception and then just logging it :)

The idea is that this code never goes to production, because some tests will fail. I think assertions can:

- help finding the source of an error (you know something goes wrong at some place, but assertions might tell you that the real reason something went wrong was that something else went wrong before this)

- help communicate under which assumptions some code was written

You are right because both things can be achieved equally well by throwing IllegalStateException (or some shorter helper method that does this) as long as performance is not an issue. My idea here was that the assert tells me that we are checking something that has already been checked anyway, but I see this is not easy to follow for others.

If however assertions are used more extensively (and possibly using some assertions that are more expensive than a simple boolean check) they can certainly be useful because they can be disabled for production. The present case is not a good example for that, but generally I think using more asserts is a good thing to do (especially in code where calculating some precondition only to find out whether or not we are in a sane state is expensive). For example RoutingKit does this a lot and its almost like you do not need tests anymore (or only a few), because when running the code with enabled assertions there is a very high chance some assertions fail when you are making a programming error and it improves readability when you are told what the expectations for the current method call were when the method was written. You are also able to easily test internal things that would never be easy to cover using tests. Ok, this is a whole other use-case than what we are talking about here, but I would just not go as far as saying that asserts are always bad or distracting.

-> I'll remove the assert :)